### ISSUE TITLE
Correlation failed, despite Redis cached DataProtectionKeys and forwarded headers

### ISSUE THREAD

**Issue by Sharor** [aspnet/Security#1844]
I was hoping to avoid making an issue, but I have run out of options and documentation to try, and [my stackoverflow post](https://stackoverflow.com/questions/51771111/c-sharp-dotnet-core-correlation-failed-behind-ingress-glbc-loadbalancer-on) has not gotten any comments either. 

I would be extremely grateful for help, as I've been stuck on this correlation failed problem for a month almost. 

My problem is as follows: 
- I am running a dotnet core 2.1 container in Kubernetes, behind a loadbalancer (GLBC). 
- I followed the [microsoft docs for web farms](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/web-farm?view=aspnetcore-2.1) as well as [proto headers forwarding for kestrel](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1). 
- I am using a Redis container to host my data protection keys, and when I executed into the container they seemed to be there. I got the idea  [here](https://github.com/aspnet/Security/issues/1755#issuecomment-392803709)
- My application needs access to Outlook and Google calendars, and I use [this excellent example for Outlook Graph integration](https://github.com/microsoftgraph/aspnetcore-connect-sample) and [Google Apis for google calendar access](https://github.com/google/google-api-dotnet-client).

Since my stackoverflow post I have tried one more thing, because I thought maybe the problem was occurring due to file persistence with the google FileStore, and so I followed [this post on how to use a database approach for storage](https://stackoverflow.com/questions/20502542/google-net-apis-any-other-datastore-other-than-the-filedatastore)  but it did not help. 

My Startup.cs is as follows: 
```
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.EntityFrameworkCore;
using Muninn.Data;
using Muninn.Models;
using Muninn.Services;
using Microsoft.AspNetCore.Authentication.Cookies;
using Microsoft.AspNetCore.Authentication.OpenIdConnect;
using Muninn.Extensions;
using Muninn.Helpers;
using Microsoft.AspNetCore.HttpOverrides;
using Microsoft.AspNetCore.DataProtection;
using Muninn.Models.ContactManagement.ContactInformations;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Rewrite;
using Muninn.Logic;
using Microsoft.AspNetCore.Http.Features;
using Google.Apis.Calendar.v3;
using Microsoft.AspNetCore.Authentication.OAuth;
using System.Security.Claims;
using Google.Apis.Auth.OAuth2.Responses;
using System;
using Google.Apis.Util.Store;
using Serilog;
using StackExchange.Redis;
using Microsoft.AspNetCore.Http;

namespace Muninn
{
    public class Startup
    {
        public IConfiguration Configuration { get; }
        public const string ObjectIdentifierType = "http://schemas.microsoft.com/identity/claims/objectidentifier";
        public const string TenantIdType = "http://schemas.microsoft.com/identity/claims/tenantid";

        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }
        
        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddDistributedRedisCache(options =>
            {
                options.Configuration = Configuration["Redis:DefaultConnection"];
                options.InstanceName = "master";
            });

            services.AddDbContext<ApplicationDbContext>(options =>
                options.UseNpgsql(Configuration.GetConnectionString("DefaultConnection")));

            services.AddIdentity<ApplicationUser, IdentityRole>()
                .AddEntityFrameworkStores<ApplicationDbContext>()
                .AddDefaultTokenProviders();

            services.AddAuthentication(sharedOptions =>
            {
                sharedOptions.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                sharedOptions.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                sharedOptions.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
            })
            .AddAzureAd(options => Configuration.Bind("AzureAd", options))
            .AddGoogle(googleOptions =>
            {
                googleOptions.Scope.Add(CalendarService.Scope.Calendar);
                googleOptions.ClientId = Configuration["Google:ClientId"];
                googleOptions.ClientSecret = Configuration["Google:ClientSecret"];
                googleOptions.AccessType = "offline";
                googleOptions.SaveTokens = true;
                googleOptions.CallbackPath = "/signin-google";
                googleOptions.Events = new OAuthEvents()
                {
                    OnCreatingTicket = async (context) =>
                    {
                        var userEmail = context.Identity.FindFirst(ClaimTypes.Email).Value;
                        Log.Information("New user logged in with Google: " + userEmail);

                        if(string.IsNullOrEmpty(context.AccessToken))
                            Log.Error("Access token was null");
                        if (string.IsNullOrEmpty(context.RefreshToken))
                            Log.Error("Refresh token was null");

                        var tokenResponse = new TokenResponse()
                        {      
                            AccessToken = context.AccessToken,
                            RefreshToken = context.RefreshToken,
                            ExpiresInSeconds = (long)context.ExpiresIn.Value.TotalSeconds,
                            IssuedUtc = DateTime.UtcNow, 
                            Scope = CalendarService.Scope.Calendar
                        };

                        var dbContext = services.BuildServiceProvider().GetService<ApplicationDbContext>();
                        IDataStore dataStore = new GoogleAuthEFDatastore(dbContext);
                        await dataStore.StoreAsync(userEmail, tokenResponse);
                        Log.Information("User has been saved to the system: " + userEmail);
                    },
                    OnRemoteFailure = async (context) =>
                    {
                        Log.Error(context.Result.Failure.Message);
                    }                    
                };
            })
            .AddCookie(options =>
            {
                options.LoginPath = "/Account/SignIn";
                options.LogoutPath = "/Account/SignOut";
            });
            services.AddMvc();


            var redis = ConnectionMultiplexer.Connect(Configuration["Redis:DefaultConnection"]);
            services.AddDataProtection()
                .PersistKeysToRedis(redis, "DataProtection-Keys")
                .SetApplicationName("myapp");           

            
            services.Configure<MvcOptions>(options =>
            {
                options.Filters.Add(new RequireHttpsAttribute());
            });
            services.Configure<CookiePolicyOptions>(options =>
            {
                options.CheckConsentNeeded = context => false;
                options.MinimumSameSitePolicy = SameSiteMode.None;
            });

            services.AddSession(options =>
            { 
                options.Cookie.Domain = ".myapp-domain.com";
                options.Cookie.SecurePolicy = Microsoft.AspNetCore.Http.CookieSecurePolicy.Always;
                options.Cookie.Name = ".MyApp.Session";
                options.IdleTimeout = TimeSpan.FromSeconds(5);
            });

            services.AddSingleton<IGraphAuthProvider, GraphAuthProvider>();
            services.AddTransient<IGraphSdkHelper, GraphSdkHelper>();
            services.AddTransient<IEmailSender, EmailSender>();
            services.AddTransient<ICalendarActions, CalendarActions>();
            services.AddTransient<IOutlookCommunication, OutlookCommunication>();
            services.Configure<GoogleAuthOptions>(Configuration.GetSection("Google"));
            services.AddTransient<IGoogleCommunication, GoogleCommunication>();
            services.Configure<FormOptions>(x =>
            {
                x.ValueLengthLimit = int.MaxValue;
                x.MultipartBodyLengthLimit = int.MaxValue; // In case of multipart
            });
            services.AddAntiforgery();

        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ApplicationDbContext dbContext)
        {
            dbContext.Database.Migrate(); 

            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseBrowserLink();
                app.UseDatabaseErrorPage();                
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
            }
            app.UseForwardedHeaders(new ForwardedHeadersOptions
            {
                ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto
            });
            app.Use(async (context, next) =>
            {
                context.Request.Scheme = "https";
                await next.Invoke();
            });
            app.UseHttpsRedirection();

            var forceSSL = new RewriteOptions()
                .AddRedirectToHttps();
            app.UseRewriter(forceSSL);

            app.UseStaticFiles();
            app.UseAuthentication();
            app.UseSession();
            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
        }
    }
}
```
My dockerfile: 

```
FROM microsoft/dotnet:2.1-sdk as builder
COPY . .
RUN dotnet restore && dotnet build
WORKDIR Muninn
RUN dotnet publish --output /app --configuration release -r ubuntu.16.04-x64

FROM microsoft/dotnet:2.1-runtime
RUN apt-get update && apt-get install -y libgdiplus
EXPOSE 5000/tcp
ENV ASPNETCORE_URLS http://*:5000
COPY --from=builder /app .
ENTRYPOINT dotnet Muninn.dll 
```

And latest stacktrace: 
```
[11:09:06 INF] Error from RemoteAuthentication: Correlation failed..
[11:09:06 ERR] An unhandled exception has occurred while executing the request.
System.NullReferenceException: Object reference not set to an instance of an object.
at Muninn.Startup.<>c.<b__6_11>d.MoveNext() in /Muninn/Startup.cs:line 103
--- End of stack trace from previous location where exception was thrown ---
at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler1.HandleRequestAsync() at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context) at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context) at Muninn.Startup.<>c.<<Configure>b__7_0>d.MoveNext() in /Muninn/Startup.cs:line 176 --- End of stack trace from previous location where exception was thrown --- at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)
[11:09:06 ERR] An unhandled exception has occurred while executing the request. 
at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)
[11:09:06 INF] Route matched with {action = "Error", controller = "Home"}. Executing action Muninn.Controllers.HomeController.Error (Muninn)
[11:09:06 INF] Executing action method Muninn.Controllers.HomeController.Error (Muninn) - Validation state: Valid
[11:09:06 INF] Executed action method Muninn.Controllers.HomeController.Error (Muninn), returned result Microsoft.AspNetCore.Mvc.ViewResult in 0.0073ms.
[11:09:06 INF] Executing ViewResult, running view Error.
[11:09:06 INF] Executed action Muninn.Controllers.HomeController.Error (Muninn) in 8.487ms
[11:09:06 ERR] An exception was thrown attempting to execute the error handler.
System.NullReferenceException: Object reference not set to an instance of an object.
at AspNetCore.Views_Shared_Error.ExecuteAsync() in /Muninn/Views/Shared/Error.cshtml:line 3
at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable1 statusCode) at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable1 statusCode)
at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultAsync(IActionResult result)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResultFilterAsyncTFilter,TFilterAsync
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResultExecutedContext context)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultFilters()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
at Microsoft.AspNetCore.Session.SessionMiddleware.Invoke(HttpContext context)
at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
at Muninn.Startup.<>c.<b__7_0>d.MoveNext() in /Muninn/Startup.cs:line 176
--- End of stack trace from previous location where exception was thrown ---
at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)
[11:09:06 ERR] Connection id "0HLG7VVB2R4E8", Request id "0HLG7VVB2R4E8:00000002": An unhandled exception was thrown by the application.
System.NullReferenceException: Object reference not set to an instance of an object.
at Muninn.Startup.<>c.<b__6_11>d.MoveNext() in /Muninn/Startup.cs:line 103
--- End of stack trace from previous location where exception was thrown ---
at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler1.HandleRequestAsync() at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context) at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context) at Muninn.Startup.<>c.<<Configure>b__7_0>d.MoveNext() in /Muninn/Startup.cs:line 176 --- End of stack trace from previous location where exception was thrown --- at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context) at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context) at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication1 application)
[11:09:06 INF] Request finished in 246.7697ms 500
```

Here I tried to log the context at *OnRemoteFailure = async (context)* which is the only difference from the stack trace in my stackoverflow posted trace. 

Any help would be greatly greatly appreciated.

**Comment by blowdart** [comment]
@Tratcher 

When are you seeing this? Is the user already logged in and trying to connect to google or outlook to get the rights? If they're already logged in then how are they logged in? With an ASP.NET identity cookie, or with an external auth cookie (for example, they just logged in with google, and now they're doing facebook).

Can you grab a fiddler trace we can look at of the flow that causes this state?

**Comment by blowdart** [comment]
Oh, darn, I had thought this might be when you login to both, one after the other.

**Comment by Sharor** [comment]
No but there is something odd going on, since when I log out of Google it takes me to the Azure endpoint - which I was gonna figure out after the correlation problem :slightly_smiling:

**Comment by Eilon** [comment]
If you do everything on a single instance (i.e. without web farm stuff), does everythign work?

**Comment by Sharor** [comment]
It works on local, with a local redis and a local db. The issue started already with just one deployed docker container however. I found that odd but figured it was due to the container being reassigned to a different node sometimes - but it seemed to happen also on a single node cluster.

**Comment by Sharor** [comment]
I'll go double check, but I'll need a day or two due to some personal things - sorry. :slightly_smiling_face: it may be that "single node" is not actually single node on GKE and I didn't verify at the time.

**Comment by Eilon** [comment]
@Sharor thanks for following up. The more info we can get about the environment in which it's failing will help us get to the bottom of this.

**Comment by Sharor** [comment]
Sorry for the delay in answering, wife went into labour :stuck_out_tongue: 

@Eilon I really appreciate the help, it's been killing me the last month :slightly_smiling_face: 
So, I've looked at GKE: 
``` 
kubectl get nodes                   
NAME                                        STATUS    ROLES     AGE       VERSION
gke-production-default-pool-2281479d-7fcp   Ready     <none>    1m        v1.9.7-gke.3
``` 

And it's only one node it gets scheduled on. 

And initially I was running one replica (git history) of my deployment, and later added it to 2. The problem was happening already on only one replica, i.e only the container on a single node in GKE. 

The structure of my cluster is quite simple and straight forward, I've a deployment (now back to 1 replica for testing) and a service and the following ingress: 

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: justdroppingby
  namespace: production
  annotations:
    kubernetes.io/ingress.class: "gce"
    ingress.kubernetes.io/rewrite-target: "/"
    kubernetes.io/ingress.allow-http: "false"
    ingress.kubernetes.io/force-ssl-redirect: "true"
spec:
  tls:
  - hosts:
    - www.myapp.com
    secretName: myapp-com-tls
  rules:
  - host: www.myapp.com
    http:
      paths:
      - path: "/*"
        backend:
          serviceName: myapp
          servicePort: 5000
```

Although the force ssl redirect/allow-http seems to not work since I can access http. 

It's always https after login though. Also a side note, my container is http but tls is terminated at this ingress - I need to figure out eventually how to get the certificate from the cluster into the container, but for now that seemed too much complexity. :slightly_smiling_face: 

Once again, really thank you for helping.

**Comment by Eilon** [comment]
@Sharor congrats!

We'll take a look at this and see what we can come up with. Just also wondering if there's any way to get a simple repro for this to try to isolate which component(s) are acting up.

**Comment by Tratcher** [comment]
A Fiddler trace and the debug output from https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1#troubleshoot are the next things to collect. We need to see the state of the failing request to understand what's going on.

**Comment by Sharor** [comment]
@Eilon thank you - you guys are being fantastic in regards to this correlation issue, hugely appreciated. 

@Tratcher 
The logging troubleshooting returned the following on all requests: 
```
Request Method: GET
Request Scheme: http
Request Path: /
Request Headers:
Connection: Keep-Alive
Via: 1.1 google
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
Cookie: .AspNetCore.Antiforgery.qrgDo4zQCUY=CfDJ8M_xM-Oxp8REm_6YDKRDeLqheNByPTykZMN2wDpXEsc_oHCY--PT-QrXJ76pLj2ZKdGFNrBtbx8ayzgy6JwsMRyks-e-eCr7aPj_vAC6aOljUbmQ4W7KRUllVH6aERCjtxEFZ3kPtwyCcX8-JzTop_U; .AspNetCore.Cookies=CfDJ8M_xM-Oxp8REm_6YDKRDeLpHC_UsiyMUhEaMpEP3b4eMoaD16TnBxSsm7FdsVgXBxOz8F1nIVc6mf9PE_wPe_8JN9hbTfzaUu6wfDuVvMCvo74EUwAK3p2qmF-hrUT1dBm6G0bmouUKgW_a38HnTlvO-yvN0iBOehNCwYNCQNyQBFmFenztkuxxHXukcDHC_d9DQz74YURl4hloHA287beZeLldG_blPrE93hE7TcBGxicwhVgAEC2EHrp1Fb6-Q6RxRqL6ZQUAWUDtt7s9NoYwZZss0U6k2wZzrHjlVsKEo0WpppJts0ZPEHVl9F4FrtCxl6DetB-WL0NWXS0FZZq_8e3jkrgf-EzWCrJtkghlyBuWulQuPLyZmtihQfNXidn08G_mdfbOf_QxBPn2uwU-BZ0s2T7-lq6waJ-G4AwBmSKKu5Fet1i5Yn942q0EVk6oKwchw1pklr5kXIxnmUuaCrwU38r9XAS2PkMXgCPfktgJP4dTowgh-lBBa4eUS3089NW5KinIChqSbMtVNxuRgJ7hwGBvbF49fNsbFWEav4GLevmHuDjs1ceCtgF8fYJEqD7xalteJhFTzDavre3Q9w00jka6Vcpc2nDlkxi61yxacTmwYOnVD784nZHc_xycC4ivsV6Z23PK58HVOzjyEoBw8MCGBOQarPsJ78AW-q5ESFskVAWe7zVX7mbou6bIDM6kAps8wFq_7gA4iy9ChlwLwYbpcQ6PIJm2NZ5BZvTbHd5kp19O7scLRMV2PY08A2z1-LeV2wR0d-P1NLX1yIaZ7W4lGcLVGfVaZfluZu5aHBFN_yK2wnupdlEWPbwdDE5JSuaM54E0Ju-CROgY0e6ygRVb-gNR4qt9axH2sO2cNDb7AJUK-lP-w4coRYFSqq2IX9CdCRpjLB7HLJ3xn3uY5Ili9rCKGPtdm7np5MyXtDkW3JguO8XMwalN7ey6kBM3wRFsZCExXHzkkHZNHey-pQcVKo469w07Pz_FTgy8jg9CPLW4rS_q2zR2lM-sagiJZJlBeufQ3eN81KZbuGF-9dzRWB2QSzbn_ZlrzRbdBHEGQSY2z-mAoVotF8PV7LjDs1UwnJPyqSFK8PDw7OfRzrqf0_tu23n5WigMJaSmriGU9I--6PD4zdxgcfKskqq_5sBhlxojpHmOZK6okY6qbQiixsImpxCAruImVTOMjxL02KsAiPDdwLm5b60ch7ZtNGlo8nfUnSrUbAlmrNBuPjYFp1QR3oVEWzsr5w0EXKQGl5tSvG5spoaYvR3r1dYYSyz4eXOnnFAixSRrqHvYS_r04jfHnuDqn2puc-EPuxsfu8yHPXjUMsTlAro21nHx2dn9eZtMGEhtNmNw
Host: www.myapp.com
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36
upgrade-insecure-requests: 1
dnt: 1
X-Cloud-Trace-Context: 1237632e61d96da6b7142ff3cc048abf/3211100041593879341
X-Forwarded-For: 212.130.123.139, 35.190.24.88
X-Forwarded-Proto: https

Request RemoteIp: ::ffff:10.4.4.1
```

As for Fiddler, apologies if this isn't the right one (never used Fiddler before) but I think this is what you want?
```

# | Result | Protocol | Host | URL | Body | Caching | Content-Type | Process | Comments | Custom
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
423 | 200 | HTTPS | console.cloud.google.com | /m/operations?authuser=1&maxResults=100 | 892 | no-cache, no-store, max-age=0, must-revalidate; Expires: Mon, 01 Jan 1990 00:00:00 GMT | application/json; charset=utf-8 | chrome:11380 |   |  


GET /m/operations?authuser=1&maxResults=100 HTTP/1.1
Host: console.cloud.google.com
Connection: keep-alive
Accept: application/json, text/plain, */*
DNT: 1
X-Goog-Request-Log-Data: clientVersion=caramel-nation-201808261502-rc00,pagePath=/kubernetes/clusters/details/:resourceZone/:resourceName,pageViewId=4252227221974175
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36
x-pan-versionid: caramel-nation-201808261502-rc00
X-Client-Data: CI22yQEIpLbJAQjEtskBCKmdygEI2J3KAQjancoBCKijygE=
Referer: https://console.cloud.google.com/kubernetes/clusters/details/europe-west1-d/production?edit=true&project=myapp-id&authuser=1&tab=details&persistent_volumes_tablesize=50&storage_class_tablesize=50&nodes_tablesize=50
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
Cookie: CONSENT=YES+DK.da+20150628-20-0; __utmc=243896023; _ga=GA1.3.1616223699.1528557407; OSID=Owbw6s7ej_NqKXEK90sgiC5s9zOeihzbR9gdsUEjeLAkO-iHqbsK9WZCMb9M86N6d2Zxjg.; S=billing-ui-v3=BsSDObCOuOV8VhwhasPylaPGH94bJ1XC:billing-ui-v3-efe=BsSDObCOuOV8VhwhasPylaPGH94bJ1XC; SID=Twbw6oMEy62-rUvF7pyWp-X9GDsQ1RaWayDMdFHl__1y0NDWRaFUGqcy7c5xDNMynqwhmg.; HSID=A00S_SQI7Pbb0GoIy; SSID=A8kfD4biHz6Si2Wky; APISID=WXxG1jqxUyyoXP-P/A0E8LKBFvho7EdjrJ; SAPISID=6uhXVWsCWdHdNsav/AFNN2Xt7cI6U3Q1-p; __utma=243896023.1616223699.1528557407.1533578990.1534320358.18; __utmz=243896023.1534320358.18.16.utmcsr=github.com|utmccn=(referral)|utmcmd=referral|utmcct=/GoogleCloudPlatform/google-cloud-dotnet; NID=137=Zgaw5-gm-7z3TsJO1QS4NMgsROUT94sDJrsPMl1genPB4O3L-H1P-uw0DTKD3u6VULj-YHiCU26-nh6gxRk-hjh1nzKQpazbcZrS0adRFlxPQtk8L4we7X_6yJvXPR2K61j7843clmfud2uJ0BGbldp1A0TxRwCayYjujkpHw26dOdIVDT3h4upDNe1f670aRajYauASC0Kq-XsA2-dux1Lrh_NqaNL0Tvbki4yVW0afF9BdELCigkqfhxq-pAR4nnak_XGFQA96QGriKr1uO_EaqE1eYXxnYZEbmMP8AEQRrGSPGg; 1P_JAR=2018-08-29-21; SIDCC=AGIhQKRl4ucQPLUkWcALtvuIuQQ__VX77qANH09-XNndJ1uKhRG05bWgxaXTfLAWWVJ-2SdHgIEh

``` 

Please just let me know if the Fiddler trace is not what you were looking for. 🙂  You guys are being amazing, really appreciate it.

**Comment by Tratcher** [comment]
You can save the fiddler session as a saz file and share the whole thing.

**Comment by Sharor** [comment]
Alright, cheers - here you are 🙂  

This session goes from : 
- Visit root
- Choose which login to use (outlook or google, here I use google) 
- Login
- Redirect (successfully, as I have not been able to make the error myself) back to root now logged in with a google (or microsoft) account. 

[totalFiddlerSession.zip](https://github.com/aspnet/Security/files/2337314/totalFiddlerSession.zip)

**Comment by Tratcher** [comment]
Ah, a trace of a successful login is only interesting if you can also capture a failed login to compare to.

**Comment by Sharor** [comment]
I'll reach out to the user who consistently gets the error and ask if I can guide him through how to get the failing trace :slightly_smiling_face: 

I haven't been able to make it fail myself which is quite frustrating.

**Comment by Sharor** [comment]
Okay, here is the same steps as before but where the correlation error happens 😉 Sorry for the additional session things, we had a google hangouts open to guide him through the process w/Fiddler

[fiddler_with_correlation_error.zip](https://github.com/aspnet/Security/files/2345279/fiddler_with_correlation_error.zip)

**Comment by Sharor** [comment]
@Tratcher I also just tested properly that indeed it is also happening with a cluster where there is one container, and one node and nothing else. 

Does it matter that the cluster is rebooted over night? I figured it shouldn't - the data protection keys are just appended, correct? 🙂 So when a new container joins, it just adds its machine key to the redis cache or reuses the existing one?

**Comment by Tratcher** [comment]
Ah, I think I see the problem. You're initiating the login flow from an HTTP endpoint but returning to an HTTPS endpoint. You need to redirect to HTTPS prior to starting your login flow. This code is wrong:
```
            app.Use(async (context, next) =>
            {
                context.Request.Scheme = "https";
                await next.Invoke();
            });
```

You're lying to the application, it thinks all requests are HTTPs but some aren't.

You start by requesting `HTTP | www.justdroppingby.com | /Account/ExternalLogin`. This returns `Set-Cookie: .AspNetCore.Correlation.Google.kev71ROMz8l_Mw5wvqZHSDIldrX7JB9RoOSt_TOS3es=N; expires=Mon, 03 Sep 2018 11:20:49 GMT; path=/signin-google; secure; httponly` (note the `secure`). It also says `redirect_uri=https%3A%2F%2Fwww.justdroppingby.com%2Fsignin-google` (note the `https`).

When the request comes back to `HTTPS | www.justdroppingby.com | /signin-google` it doesn't have any cookies and it fails.

**Comment by Sharor** [comment]
@Tratcher  Oh! What a fantastic finding, I did not even consider that Kestrel would look at its own request in that way. 

Do you have some documentation you can direct me to that would solve this problem or does it need to be sorted out on loadbalancer level? 

I would like to solve it both on application level (ie not allow http traffic, which I thought these lines in y startup were already doing: 
```
app.UseHttpsRedirection();
 
var forceSSL = new RewriteOptions()
    .AddRedirectToHttps();
app.UseRewriter(forceSSL);
```

and also on load balancer level, but I'll sort that out straight way. 😉

**Comment by Tratcher** [comment]
Start here:
https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1
https://docs.microsoft.com/en-us/aspnet/core/security/enforcing-ssl?view=aspnetcore-2.1&tabs=visual-studio

**Comment by Sharor** [comment]
@Tratcher So I have finally solved it with your help, thank you so very much your help has been invaluable. Also your work/issues left around this repo and other places helped a lot, truly appreciated before I made this issue. 

The full solution in the end was a bit different then I thought, but nonetheless: 
- Enable https only (redirect http to s) on loadbalancer level, and keep container (.NET core) running purely http already solves it. 
- The correct longterm solution is to then also amount the certificate into the container and enable https on the container level but seemed not necessary to make things work. 

So to summarize for anyone coming later with the same issue: 
- The issue happens because the application thinks all traffic is https, but not enforcing https before reaching the container causes the issue. 
- The solution is redirect at loadbalancer level. 
- The proper solution is to also redirect at container level. 

Cheers! :wink:
